File: Cousins.py
#Time Complexity: O(n)
#Space Complexity: O(h)
# Cousins in a tree are at same level but have different parents. If they have same parent then they are siblings not cousins

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        self.x_parent = None
        self.y_parent = None
        self.x_height = 0
        self.y_height = 0
        self.DFS(root, 0, x, y,None)
        if self.x_parent and self.y_parent and self.x_parent != self.y_parent and self.x_height == self.y_height:
            return True
        return False

    def DFS(self, root, depth, x, y, parent):
        # Base case
        if root is None:
            return

        # Logic
        if root.val==x:
            self.x_height=depth
            self.x_parent=parent
        elif root.val==y:
            self.y_height=depth
            self.y_parent=parent
        if root.left:
            self.DFS(root.left, depth+1, x, y, root)
        if root.right:
            self.DFS(root.right, depth+1, x, y, root)

File: Cousins.cpp
// TC: O(N), SC:O(2^H):

// BFS solution
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
 
        queue<TreeNode*>q;
        q.push(root);

        while(!q.empty())
        {
            int s=q.size();
            bool xflag=false;
            bool yflag=false;
           
           
            for(int i=0;i<s;i++)
            {
                TreeNode* a=q.front();
                q.pop();
                // checking common parent conditions
                if((a->left && a->left->val==x)&&(a->right && a->right->val==y))
                {
                    return false;
                }

                if((a->left && a->left->val==y)&&(a->right && a->right->val==x))
                {
                    return false;
                }

                if(a->val==x)
                xflag=true;

                if(a->val==y )
                yflag=true;


                if(a->left!=nullptr)
                q.push(a->left);

                if(a->right!=nullptr)
                q.push(a->right);

         
            }
            if(xflag==true && yflag==true)
            return true;
            if(xflag==true || yflag==true)
            return false;
    
         
        }

        return true;
    }
};

TC:O(N)
SC:O(2^H)
// DFS solution

class Solution {
public:
    TreeNode* xpar;
    TreeNode* ypar;
    int xlvl,ylvl;
    bool isCousins(TreeNode* root, int x, int y) 
    {
        if(root==nullptr)
         return false;

        cousins(root,nullptr,x,y,0);

        if(xpar != ypar && xlvl == ylvl)
        {
           return true;
        }
        return false;
    }

    void cousins(TreeNode* root,TreeNode* par,int x,int y,int lvl)
    {
            if(root==nullptr)
                return;
            
            if(root->val==x)
            {
                xpar=par;
                xlvl=lvl;
            }

            if(root->val==y)
            {
                ypar=par;
                ylvl=lvl;
            }

            cousins(root->left,root,x,y,lvl+1);
            cousins(root->right,root,x,y,lvl+1);

         
    }
}
